[WTF] Convert VectorTypeOperations to use constexpr if

fujii opened the pull request:
#### 97c0170f3a51701d52fdee3d9fbdaee24f6b7a57
<pre>
[WTF] Convert VectorTypeOperations to use constexpr if
<a href="https://bugs.webkit.org/show_bug.cgi?id=299435">https://bugs.webkit.org/show_bug.cgi?id=299435</a>

Reviewed by Darin Adler.

Modernized VectorTypeOperations to use constexpr if rather than the
old template metaprogramming style.

Renamed a variable `cur` to `current`.

* Source/WTF/wtf/Vector.h:
(WTF::VectorCopier::uninitializedCopy):
(WTF::VectorTypeOperations::destruct):
(WTF::VectorTypeOperations::initializeIfNonPOD):
(WTF::VectorTypeOperations::initialize):
(WTF::VectorTypeOperations::initializeWithArgs):
(WTF::VectorTypeOperations::move):
(WTF::VectorTypeOperations::moveOverlapping):
(WTF::VectorTypeOperations::uninitializedCopy):
(WTF::VectorTypeOperations::uninitializedFill):
(WTF::VectorTypeOperations::compare):
(WTF::Vector::Vector):
(WTF::Malloc&gt;::append):
(WTF::Malloc&gt;::unsafeAppendWithoutCapacityCheck):
(WTF::Malloc&gt;::insertSpan):

Canonical link: <a href="https://commits.webkit.org/300661@main">https://commits.webkit.org/300661@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e21fc77e51c5abf2b7c78b1924409cd47242b401

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123385 "7 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/43100 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/33796 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/130099 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/75509 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/43823 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/51694 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/93791 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/62230 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126338 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/34920 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110393 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74418 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/33888 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/28551 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/73616 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/115544 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104634 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/28777 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/132816 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/121916 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50335 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38310 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/102281 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/50711 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106616 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/102134 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47491 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/25714 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/47125 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19435 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50190 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/55951 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/152271 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/49663 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/38923 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53011 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/51339 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/e43c0d3c0da2f9c3060b63ad396c8b6d4d866dcd. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51253

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/e6f0d531a1a0ac61cbf375218e982ce1e46c52e3. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51253

darinadler commented on Source/WTF/wtf/Vector.h:
```diff
@@ -108,217 +94,113 @@ struct VectorInitializer<true, false, T>
+    }
+
+    template<typename U>
+    static void uninitializedImport(const U* src, const U* srcEnd, T* dst)
```
Not sure why we are introducing new terminology, "import". Why isn't this just `uninitializedCopy`?

fujii replied:
I restore VectorCopier for this case.

fujii replied:
Fixed.

darinadler commented on Source/WTF/wtf/Vector.h:
```diff
@@ -108,217 +94,113 @@ struct VectorInitializer<true, false, T>
+    template<typename U>
+    static void uninitializedImport(const U* src, const U* srcEnd, T* dst)
+    {
+        static_assert(!std::is_const_v<T>);
```
Why is this needed?

fujii replied:
Removed.

darinadler commented on Source/WTF/wtf/Vector.h:
```diff
@@ -108,217 +94,113 @@ struct VectorInitializer<true, false, T>
+    {
+        static_assert(!std::is_const_v<T>);
+        if constexpr (std::is_trivially_copyable_v<T> && std::is_same_v<T, U>)
+            memcpy(static_cast<void*>(dst), static_cast<void*>(const_cast<T*>(src)), reinterpret_cast<const char*>(srcEnd) - reinterpret_cast<const char*>(src));
```
I don't think we need fhe `static_cast<void*>`. What compiler error do we get if we omit those? Please remove them.

We do not need `const_cast`. Please remove it.

darinadler commented on Source/WTF/wtf/Vector.h:
```diff
@@ -108,217 +94,113 @@ struct VectorInitializer<true, false, T>
+    static void uninitializedImport(const U* src, const U* srcEnd, T* dst)
+    {
+        static_assert(!std::is_const_v<T>);
+        if constexpr (std::is_trivially_copyable_v<T> && std::is_same_v<T, U>)
```
Seems like this won't ever be true because `U` is const and `T` is not. Are you sure the optimization is ever happening? How did you check that?

The old code checked `is_trivial` and this new code checks `is_trivially_copyable. Is this new check better? How did we verify it` doesn't remove an optimization or optimize incorrectly?

darinadler replied:
> `U` is non-const.

I see. I guess I read the text wrong.

> `std::is_trivial` is deprecated in C++26.

That's interesting and it's forward looking to not sue something that will be deprecated in the future.

Does `is_trivially_copyable` give the exact sakes behavior or are there cases it will handle differently?

darinadler commented on Source/WTF/wtf/Vector.h:
```diff
@@ -108,217 +94,113 @@ struct VectorInitializer<true, false, T>
     {
-        VectorFiller<VectorTraits<T>::canFillWithMemset, T>::uninitializedFill(dst, dstEnd, val);
+        if constexpr (VectorTraits<T>::canFillWithMemset) {
+            static_assert(sizeof(T) == 1, "Size of type T should be equal to one!");
```
This code is so old that when it was written I think `static_assert` required a comment. But now it doesn't and this comment says exactly what the expression does and adds nothing. Please remove it.

fujii replied:
Fixed.

darinadler commented on Source/WTF/wtf/Vector.h:
```diff
@@ -56,48 +56,34 @@ DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER(VectorBuffer);
+struct VectorTypeOperations
 {
-    static void destruct(T* begin, T* end) 
+    static void destruct(T* begin, T* end)
```
I don't think we need this at all. I think callers can just used `std::destroy` or `std::ranges::destroy` or if we still want this function for some reason we could use that here instead of writing a loop.

darinadler commented on Source/WTF/wtf/Vector.h:
```diff
@@ -56,48 +56,34 @@ DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER(VectorBuffer);
-    static void initializeIfNonPOD(T*, T*) { }
-
-    static void initialize(T* begin, T* end)
+    static void destruct(std::span<T> span)
```
I think callers can just use `std::ranges::destroy` instead.

darinadler replied:
> It sounds a nice idea to use `std::ranges`. But, I don't want to do that in this change. Too risky.

Maybe we should off on this whole change. The entire thing is just "better style"; what is the primary benefit to the change? Does it fix any bugs?

fujii replied:
https://commits.webkit.org/276220@main introduced the performance regression for `WTF::Vector`.
I think one of the reason is TMP is too dificult. That's the motivation of this change.
https://bugs.webkit.org/show_bug.cgi?id=299008

darinadler commented on Source/WTF/wtf/Vector.h:
```diff
@@ -108,217 +94,113 @@ struct VectorInitializer<true, false, T>
     template<typename... Args>
     static void initializeWithArgs(T* begin, T* end, Args&&... args)
     {
+        static_assert(VectorTraits<T>::needsInitialization);
+        static_assert(!VectorTraits<T>::canInitializeWithMemset);
```
Why add these assertions? This seems an unnecessary new restriction. Maybe a caller chooses to initialize something with arguments that doesn't need initialization or that can be initialized with `memset`? Why wouldn't we support that?

fujii replied:
I created a test case that demonstrates the problem of the original code. WebKit doesn't have code for that case because it doesn't need it yet.
https://gist.github.com/fujii/7c2d49f351dcf26100211d7e595c1b48

darinadler commented on Source/WTF/wtf/Vector.h:
```diff
@@ -56,48 +56,34 @@ DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER(VectorBuffer);
+            if constexpr (VectorTraits<T>::canInitializeWithMemset)
+                memset(static_cast<void*>(begin), 0, reinterpret_cast<char*>(end) - reinterpret_cast<char*>(begin));
+            else {
+                for (T* cur = begin; cur != end; ++cur)
+                    new (NotNull, cur) T();
```
In new code we would follow WebKit coding style and use a word rather than `cur`.

I think we can use `std::uninitialized_default_construct` or `std::ranges::uninitialized_default_construct` instead of writing a loop, although we will need to verify that it's still equally efficient.

darinadler commented on Source/WTF/wtf/Vector.h:
```diff
@@ -56,48 +56,34 @@ DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER(VectorBuffer);
-template<typename T>
-struct VectorInitializer<true, false, T>
-{
     static void initializeIfNonPOD(T* begin, T* end)
```
I don't understand why this is still a separate function from `initialize` since `initialize` just calls it unconditionally. How does this work?

darinadler commented on Source/WTF/wtf/Vector.h:
```diff
@@ -108,217 +94,113 @@ struct VectorInitializer<true, false, T>
-                --srcEnd;
-                --dstEnd;
-                new (NotNull, dstEnd) T(WTFMove(*srcEnd));
-                srcEnd->~T();
+                new (NotNull, dst) T(WTFMove(*src));
+                src->~T();
+                ++dst;
+                ++src;
             }
```
I think we can use `std::uninitialized_copy` or `std::ranges_uninitialized_copy` here instead of writing our own loop, although we will need to verify that it's still equally efficient.

darinadler commented on Source/WTF/wtf/Vector.h:
```diff
@@ -108,217 +94,113 @@ struct VectorInitializer<true, false, T>
+        static_assert(VectorTraits<T>::needsInitialization);
+        static_assert(!VectorTraits<T>::canInitializeWithMemset);
+
         for (T *cur = begin; cur != end; ++cur)
             new (NotNull, cur) T(args...);
```
I think we can use `std::uninitialized_value_construct` or `std::ranges_uninitialized_value_construct` here instead of writing our own loop, although we will need to verify that it's still equally efficient.

darinadler commented on Source/WTF/wtf/Vector.h:
```diff
@@ -108,217 +94,113 @@ struct VectorInitializer<true, false, T>
+        if constexpr (VectorTraits<T>::canMoveWithMemcpy)
+            memmove(static_cast<void*>(dst), static_cast<void*>(const_cast<T*>(src)), reinterpret_cast<const char*>(srcEnd) - reinterpret_cast<const char*>(src));
+        else {
+            if (src > dst)
+                move(src, srcEnd, dst);
+            else {
+                T* dstEnd = dst + (srcEnd - src);
+                while (src != srcEnd) {
+                    --srcEnd;
+                    --dstEnd;
+                    new (NotNull, dstEnd) T(WTFMove(*srcEnd));
+                    srcEnd->~T();
+                }
+            }
```
I think we can use `std::uninitialized_move` or `std::ranges_uninitialized_move` here instead of writing our own loop, although we will need to verify that it's still equally efficient.

darinadler commented on Source/WTF/wtf/Vector.h:
```diff
@@ -108,217 +94,113 @@ struct VectorInitializer<true, false, T>
+        if constexpr (VectorTraits<T>::canCopyWithMemcpy)
+            memcpy(static_cast<void*>(dst), static_cast<void*>(const_cast<T*>(src)), reinterpret_cast<const char*>(srcEnd) - reinterpret_cast<const char*>(src));
+        else {
+            while (src != srcEnd) {
+                new (NotNull, dst) T(*src);
+                ++dst;
+                ++src;
+            }
```
I think we can use `std::uninitialized_copy` or `std::ranges_uninitialized_copy` here instead of writing our own loop, although we will need to verify that it's still equally efficient.

fujii replied:
This code is old which has a lot of problems. I'm focusing on TMP in this change. out of scope of this change.

darinadler commented on Source/WTF/wtf/Vector.h:
```diff
@@ -108,217 +94,113 @@ struct VectorInitializer<true, false, T>
+            static_assert(sizeof(T) == 1, "Size of type T should be equal to one!");
+            memset(dst, val, dstEnd - dst);
+        } else {
+            while (dst != dstEnd) {
+                new (NotNull, dst) T(val);
+                ++dst;
+            }
```
I think we can use `std::uninitialized_fill` or `std::ranges_uninitialized_fill` here instead of writing our own loop, although we will need to verify that it's still equally efficient.

darinadler commented on Source/WTF/wtf/Vector.h:
```diff
@@ -108,217 +94,113 @@ struct VectorInitializer<true, false, T>
-        if (src > dst)
-            move(src, srcEnd, dst);
+        if constexpr (VectorTraits<T>::canMoveWithMemcpy)
+            memcpy(static_cast<void*>(dst), static_cast<void*>(const_cast<T*>(src)), reinterpret_cast<const char*>(srcEnd) - reinterpret_cast<const char*>(src));
```
> a non-const pointer can be casted to a const pointer. I think that's the reason casting to `const char*`.

We should be casting to `char*` instead.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/9ee4030ad6d2ed7c76a6f7fa79562b7363ca948b. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51253

fujii commented on Source/WTF/wtf/Vector.h:
```diff
@@ -56,48 +56,34 @@ DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER(VectorBuffer);
     {
-        for (T* cur = begin; cur != end; ++cur)
-            cur->~T();
+        if constexpr (!std::is_trivially_destructible_v<T>) {
+            for (T* cur = begin; cur != end; ++cur)
+                cur->~T();
+        }
```
I'd like to keep it untouched in this change.

darinadler commented on Source/WTF/wtf/Vector.h:
```diff
@@ -56,269 +56,152 @@ DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER(VectorBuffer);
     {
-        if (src > dst)
-            move(src, srcEnd, dst);
+        if constexpr (std::is_trivial_v<T> && std::is_same_v<T, U>)
```
Should we consider `std::same_as` instead of `std::is_same_v`? Kind of annoying that C++ has both.

darinadler commented on Source/WTF/wtf/Vector.h:
```diff
@@ -56,269 +56,152 @@ DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER(VectorBuffer);
-        if (src > dst)
-            move(src, srcEnd, dst);
+        if constexpr (std::is_trivial_v<T> && std::is_same_v<T, U>)
+            memcpy(static_cast<void*>(dst), static_cast<void*>(const_cast<T*>(src)), reinterpret_cast<const char*>(srcEnd) - reinterpret_cast<const char*>(src));
```
What was our conclusion about the need for `static_cast<void*>` and `const_cast`? Did we try removing it and run into problems?

darinadler commented on Source/WTF/wtf/Vector.h:
```diff
@@ -56,269 +56,152 @@ DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER(VectorBuffer);
     {
-        VectorInitializer<VectorTraits<T>::needsInitialization, VectorTraits<T>::canInitializeWithMemset, T>::initialize(begin, end);
+        if constexpr (VectorTraits<T>::canInitializeWithMemset)
+            memset(static_cast<void*>(begin), 0, reinterpret_cast<char*>(end) - reinterpret_cast<char*>(begin));
```
Same question about this `static_cast<void*>`.

darinadler commented on Source/WTF/wtf/Vector.h:
```diff
@@ -56,269 +56,152 @@ DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER(VectorBuffer);
-        VectorFiller<VectorTraits<T>::canFillWithMemset, T>::uninitializedFill(dst, dstEnd, val);
+        if constexpr (VectorTraits<T>::canFillWithMemset) {
+            static_assert(sizeof(T) == 1);
+            memset(dst, val, dstEnd - dst);
```
Seems like this will do something strange if `val` isn’t a scalar that converts to `int` without loss. I can imagine single bytes that when you convert them to `int` lose something. I’m surprised with all the other `static_assert` that we aren’t doing some assertions about that here. I guess it’s nothing really to worry about.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/e21fc77e51c5abf2b7c78b1924409cd47242b401. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51253

fujii committed:
[WTF] Convert VectorTypeOperations to use constexpr if
https://bugs.webkit.org/show_bug.cgi?id=299435

Reviewed by Darin Adler.

Modernized VectorTypeOperations to use constexpr if rather than the
old template metaprogramming style.

Renamed a variable `cur` to `current`.

* Source/WTF/wtf/Vector.h:
(WTF::VectorCopier::uninitializedCopy):
(WTF::VectorTypeOperations::destruct):
(WTF::VectorTypeOperations::initializeIfNonPOD):
(WTF::VectorTypeOperations::initialize):
(WTF::VectorTypeOperations::initializeWithArgs):
(WTF::VectorTypeOperations::move):
(WTF::VectorTypeOperations::moveOverlapping):
(WTF::VectorTypeOperations::uninitializedCopy):
(WTF::VectorTypeOperations::uninitializedFill):
(WTF::VectorTypeOperations::compare):
(WTF::Vector::Vector):
(WTF::Malloc>::append):
(WTF::Malloc>::unsafeAppendWithoutCapacityCheck):
(WTF::Malloc>::insertSpan):

Canonical link: https://commits.webkit.org/300661@main
